src/search.js

```diff
@@ -0,0 +1,58 @@
+(function() {
+
+  'use strict';
+
+  /* global require */
+
+  var HMS = require('hindley-milner-search');
+
+  //  initSigNodes :: Node -> [Node]
+  function initSigNodes(resultsContainer) {
+    return Array.prototype.reduce.call(
+      document.getElementById('toc').querySelectorAll('li a code'),
+      function(r, _el) {
+        var el = document.createElement('div');
+        el.appendChild(_el.parentNode.cloneNode(true));
+        el.className = 'search-result';
+        var k = el.childNodes[0].getAttribute('href');
+        if (!Object.prototype.hasOwnProperty.call(r.set, k)) {
+          r.set[k] = true;
+          r.uniqs.push(el);
+        }
+        return r;
+      },
+      {set: {}, uniqs: []}
+    ).uniqs;
+  }
+
+  window.addEventListener('DOMContentLoaded', function() {
+    var $container = document.getElementById('yes-matches');
+    var $containerNo = document.getElementById('no-matches');
+    var $els = initSigNodes(document.getElementById('search-results'));
+    var search = HMS.init(Array.prototype.map.call($els, function(el) {
+      return el.innerText.trim().split(/\s+/).join(' ');
+    }), {fuzzy: false}).search;
+    var $prevSearch = '';
+    var searchInput = document.getElementById('search-input');
+    searchInput.focus();
+    searchInput.addEventListener('keyup', function(event) {
+      var q = event.target.value;
+      if (q !== $prevSearch) {
+        $prevSearch = q;
+        var matches = search(q);
+        $containerNo.style.display =
+          matches.length === 0 && q !== '' ? 'block' : 'none';
+        while ($container.childNodes.length > 0) {
+          $container.removeChild($container.childNodes[0]);
+        }
+        // if there is no query, do not show any matches
+        if (q !== '') {
```

Comment:
Since we'll be displaying results at __/search/__ rather than at __/__, displaying all results for an empty query seems right to me. I've removed this check. :)